### PR TITLE
fix badger double open on daemon --import-snapshot; chainstore lifecycle

### LINKED ISSUE #4428: dangling `reorgWorker` after snapshot import
https://github.com/filecoin-project/lotus/blob/151577f04ef264a63bedb86ef6bf527a64743c5d/chain/store/store.go#L190

Normally not an issue because the repo/store is not closed until the node stops but when importing we open/close the repo and the dangling thread fails with `datastore closed` error.

### PR DESCRIPTION
**Pull request by raulk**
This PR fixes https://github.com/filecoin-project/lotus/issues/4850. The problem was that the IdStore did not implement `io.Closer`, therefore the store remained open after the import. This required a release of go-ipfs-blockstore, which has now been integrated here.

I also noticed this error (which I think has been there since forever; EDIT: confirmed, this was already there! #4428) just after the import is done, and before we start the node:

```
2020-11-16T15:08:17.027Z        INFO    main    lotus/daemon.go:474     accepting [bafy2bzaceckovkpb5vua5s2sbltjg2lynikeehepztyjsf7n2nrehot36bwdc] as new head
2020-11-16T15:08:17.027Z        INFO    chainstore      store/store.go:500      New heaviest tipset! [bafy2bzaceckovkpb5vua5s2sbltjg2lynikeehepztyjsf7n2nrehot36bwdc] (height=240581)
2020-11-16T15:08:17.027Z        INFO    chainstore      store/store.go:687      failed to fetch right.Parents: get block bafy2bzaceaoyzhu6i2qpsi2efx7wp2z7kta54typ5fewd52mlsvg4azjrnkjg: badger blockstore closed
2020-11-16T15:08:17.027Z        ERROR   chainstore      store/store.go:420      computing reorg ops failed: get block bafy2bzaceaoyzhu6i2qpsi2efx7wp2z7kta54typ5fewd52mlsvg4azjrnkjg: badger blockstore closed
```

This happens because the `SetHead` inside `ImportChain` triggers a reorg, but by the team the event arrives to the event loop, the badger store is closed. There are two problems here:

1. The chainstore does not accept a context, thus it the reorg runner has no way to stop. Currently, this would leak a goroutine. Fixes #4428.
2. This forced `SetHead` should not trigger a reorg; we can use the newly-added `ForceHeadSilent` operation to set the head without a reorg.

This PR fixes both those things too.

### CODE REVIEW

**Review by magik6k — APPROVED** [review]

File: chain/store/store.go [hunk 1]
```diff
@@ -383,7 +382,7 @@ func (cs *ChainStore) MaybeTakeHeavierTipSet(ctx context.Context, ts *types.TipS
 // particular tipset to carry out a benchmark, verification, etc. on a chain
 // segment.
 func (cs *ChainStore) ForceHeadSilent(_ context.Context, ts *types.TipSet) error {
-	log.Warnf("(!!!) forcing a new head silently; only use this only for testing; new head: %s", ts)
```
Comment (whyrusleeping) [inline chain/store/store.go:386]: eh?

**Comment by Kubuxu** [comment]
Personally I would prefer if chainstore had a Close method instead of relying on the context for the lifetime.

File: chain/store/store.go (hunk 1 above)
Comment (raulk) [reply chain/store/store.go:386]: ?

**Comment by raulk** [comment]
@Kubuxu agree; addressed in 0c6072a.

**Review by Kubuxu — APPROVED** [review]

File: chain/store/store.go (hunk 1 above)
Comment (magik6k) [reply chain/store/store.go:386]: There was no valid use for this outside testing before. When importing snapshot it doesn't make sense to emit reorg notifications as nothing is listening on that chainstore, and everything that cares about chain state does check current state when subscribing

tldr this is fine

**Comment by raulk** [comment]
@magik6k wanna merge this?

**Review by magik6k — APPROVED** [review]
